vault/hostinfo.go

```diff
@@ -0,0 +1,68 @@
+package vault
+
+import (
+	"runtime"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/disk"
+	"github.com/shirou/gopsutil/host"
+	"github.com/shirou/gopsutil/mem"
+)
+
+type HostInfo struct {
+	CollectionTime time.Time              `json:"collection_time"`
+	CPU            []cpu.InfoStat         `json:"cpu"`
+	Disk           *disk.UsageStat        `json:"disk"`
+	Host           *host.InfoStat         `json:"host"`
+	Memory         *mem.VirtualMemoryStat `json:"memory"`
+}
+
+type HostInfoError struct {
+	Err error
+}
+
+func (e *HostInfoError) WrappedErrors() []error {
+	return []error{e.Err}
+}
+
+func (e *HostInfoError) Error() string {
+	return e.Err.Error()
+}
+
+func (c *Core) CollectHostInfo() (*HostInfo, error) {
+	var retErr *multierror.Error
+	info := &HostInfo{CollectionTime: time.Now()}
+
+	if h, err := host.Info(); err != nil {
+		retErr = multierror.Append(retErr, &HostInfoError{err})
+	} else {
+		info.Host = h
+	}
+
+	if v, err := mem.VirtualMemory(); err != nil {
+		retErr = multierror.Append(retErr, &HostInfoError{err})
+	} else {
+		info.Memory = v
+	}
+
+	diskPath := "/"
+	if runtime.GOOS == "windows" {
+		diskPath = "C:"
+	}
+
+	if d, err := disk.Usage(diskPath); err != nil {
+		retErr = multierror.Append(retErr, &HostInfoError{err})
+	} else {
+		info.Disk = d
+	}
+
+	if c, err := cpu.Info(); err != nil {
+		retErr = multierror.Append(retErr, &HostInfoError{err})
+	} else {
+		info.CPU = c
+	}
+
+	return info, retErr.ErrorOrNil()
```

Comment:
I am returning a `multierror.Error` in here vs returning error whenever a single call to fetch information fails. This allows us to return partial host information even when we're unable to get certain bits.

These errors [gets translated](https://github.com/hashicorp/vault/pull/7330/files#diff-a145099c8e3d917858619dcfb2ae09b9R2623) to response warnings in `handleHostInfo` at the logical layer. However, for the client API, these warnings are swallowed, so I am not sure if I should include them as `Warnings` in `api.HostInfoResponse`.

Reply:
This is no longer a concern since the endpoint is queried via `Logical().Read(...)`, and the warnings are preserved under `Secret.Warnings`.